lib/xapi-stdext-unix/unixext.ml

```diff
@@ -430,48 +425,33 @@ let rec really_read fd string off n =
     really_read fd string (off+m) (n-m)
 
 let really_read_string fd length =
-  let buf = String.make length '\000' in
+  let buf = Bytes.make length '\000' in
   really_read fd buf 0 length;
-  buf
+  Bytes.unsafe_to_string buf
 
 let try_read_string ?limit fd =
   let buf = Buffer.create 0 in
   let chunk = match limit with None -> 4096 | Some x -> x in
-  let cache = String.make chunk '\000' in
+  let cache = Bytes.make chunk '\000' in
   let finished = ref false in
   while not !finished do
     let to_read = match limit with
       | Some x -> min (x - (Buffer.length buf)) chunk
       | None -> chunk in
     let read_bytes = Unix.read fd cache 0 to_read in
-    Buffer.add_substring buf cache 0 read_bytes;
+    Buffer.add_subbytes buf cache 0 read_bytes;
     if read_bytes = 0 then finished := true
   done;
   Buffer.contents buf
 
-let really_read_bigbuffer fd bigbuf n =
-  let chunk = 4096 in
-  let s = String.make chunk '\000' in
-  let written = ref 0L in
-  while !written < n do
-    let remaining = Int64.sub n !written in
-    let to_write = min remaining (Int64.of_int chunk) in
-    really_read fd s 0 (Int64.to_int to_write);
-    Bigbuffer.append_substring bigbuf s 0 (Int64.to_int to_write);
-    written := Int64.add !written to_write;
-  done
-
+(* This was equivalent to Unix.write - deprecating *)
 let really_write fd string off n =
-  let written = ref 0 in
-  while !written < n
-  do
-    let wr = Unix.write fd string (off + !written) (n - !written) in
-    written := wr + !written
-  done
+  Unix.write fd string off n |> ignore
 
 (* Ideally, really_write would be implemented with optional arguments ?(off=0) ?(len=String.length string) *)
 let really_write_string fd string =
-  really_write fd string 0 (String.length string)
+  let payload = Bytes.unsafe_of_string string in
```

Comment:
Again, I believe Unix.write may need several writes.

Reply:
No, it already does the multiple writes. You need to use `Unix.write_single` otherwise

Reply:
Cool. I did not know that `Unix.write` implements the looping itself. The name suggested it would do a write(2).